staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -295,20 +295,11 @@ func (qs *queueSet) StartRequest(ctx context.Context, width *fqrequest.Width, ha
 	configName := qs.qCfg.Name
 
 	if doneCh != nil {
-		qs.preCreateOrUnblockGoroutine()
 		go func() {
 			defer runtime.HandleCrash()
-			qs.goroutineDoneOrBlocked()
```

Comment:
One more thing I want to do in subsequent PR is to ensure that production code isn't really tracking goroutines at all (it doesn't use this information) by using non-tracking promise in prod and tracking one (build on top of non-tracking one) in tests.

Reply:
The production code already is not really tracking goroutines --- see https://github.com/kubernetes/apiserver/blob/b94b411ffc0d129451a850a78bd39b9457a15518/pkg/util/flowcontrol/apf_filter.go#L85

By removing the updates through the counter interface, this makes it impossible to test queueset with a fake clock.  It is important to be able to test with a fake clock:
- so that we can control timing precisely, and
- so that we can run tests faster.

Reply:
Sorry - I wasn't clear exactly.
I'm aware we're using no-op tracker in production code.

But I'm really not a fan of having test-only related code in production code. This complicates production code.
All the test-related code should really be separate out from the code that is running in production.

> By removing the updates through the counter interface, this makes it impossible to test queueset with a fake clock. It is important to be able to test with a fake clock:

I disagree with this one. Currently none of the tests is exercising this. But even if at some point we extend our testing to take advantage of it, what we should do is:
- make a class function from this goroutine [and leave it without counter]
- in tests, define our own `TestingQueuset` (or whatever name we choose) that will be basically inheriting everything from the QueueSet, we will just redefing this single function

I can live with complicate tests, but I really want to have production code as simple as possible [it's not just about P&F of course].

Reply:
https://github.com/kubernetes/apiserver/blob/b94b411ffc0d129451a850a78bd39b9457a15518/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go#L653 _does_ exercise this.

Reply:
If we factor out this goroutine, the resulting surface seems odder to me than the one we have here.  What is the meaning of this goroutine?  It's just a random bit of implementation.  Counting active goroutines, on the other hand, is a simple self-contained semantic.

Reply:
> https://github.com/kubernetes/apiserver/blob/b94b411ffc0d129451a850a78bd39b9457a15518/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go#L653 does exercise this.

It exercises that code path, but it doesn't exercise the counter:
- clk is not used anywhere after initialization
- counter is only updated through the text but never check
[as the test shows - it passed consistently after removing it]

> If we factor out this goroutine plus old line 298, the resulting surface seems odder (less simple) to me than the one we have now. What is the meaning of this goroutine plus old line 298? It's just a random bit of implementation. Counting active goroutines, on the other hand, is a simple self-contained semantic.

I disagree with this. When I'm reading the code I would like to be able to quickly understand what exactly is happening in production. And mixing test code with production code and just initializing it differently is not the best practice.